libs/vkd3d/resource.c

```diff
@@ -1960,6 +1960,31 @@ struct d3d12_resource *unsafe_impl_from_ID3D12Resource(ID3D12Resource *iface)
     return unsafe_impl_from_ID3D12Resource1((ID3D12Resource1 *)iface);
 }
 
+VkImageSubresource d3d12_resource_get_vk_subresource(const struct d3d12_resource *resource, uint32_t subresource_idx, bool all_aspects)
+{
+    const struct vkd3d_format *format = vkd3d_format_from_d3d12_resource_desc(resource->device, &resource->desc, 0);
+    VkImageSubresource subresource;
+
+    subresource.aspectMask = format->vk_aspect_mask;
+    subresource.mipLevel = subresource_idx % resource->desc.MipLevels;
+    subresource.arrayLayer = (subresource_idx / resource->desc.MipLevels) % resource->desc.DepthOrArraySize;
+
+    if (!all_aspects)
+    {
+        uint32_t i, plane_idx = subresource_idx / (resource->desc.MipLevels * resource->desc.DepthOrArraySize);
+
+        /* Clear the lower 'plane_idx' bits, then isolate the lowest bit.
+         * This works because aspect numbering between D3D12 and Vulkan
+         * is the same. */
+        for (i = 0; i < plane_idx; i++)
```

Comment:
It is a bit hard to understand (without knowing the full implementation) why this is done.

Reply:
Changed the comment.

Reply:
Ok, I think I get it now. Clever bithack indeed.